File: samples/c/pci_dump/pci_dump.c
/* Jungo Connectivity Confidential. Copyright (c) 2022 Jungo Connectivity Ltd.  https://www.jungo.com */

////////////////////////////////////////////////////////////////
// File - PCI_DUMP.C
//
// A utility for getting a dump of all the PCI configuration
// registers of the PCI cards installed.
//
// Note: This code sample is provided AS-IS and as a guiding sample only.
////////////////////////////////////////////////////////////////

#include "windrvr.h"
#include "wdc_diag_lib.h"
#include "status_strings.h"

#if defined(LINUX)
#include <stdarg.h>
#endif

int main(int argc, char *argv[])
{

    int argi = 1;
    WD_PCI_SLOT pciSlot = {0};
    BOOL single = FALSE;
    DWORD dwStatus = WD_WINDRIVER_STATUS_ERROR;
    FILE *fp = stdout;

    if (argc !=1 && argc != 2 && argc != 4 && argc != 5)
    {
        printf("USAGE: %s [filename] [bus# slot# function#]\n", argv[0]);
        return -1;
    }
    if (argc == 2 || argc == 5)
    {
        if (NULL == (fp = fopen(argv[argi++], "w")))
        {
            perror(argv[1]);
            goto Exit;
        }
    }
    if (argc >= 4)
    {
        sscanf(argv[argi++], "%ld", (long *)&pciSlot.dwBus);
        sscanf(argv[argi++], "%ld", (long *)&pciSlot.dwSlot);
        sscanf(argv[argi++], "%ld", (long *)&pciSlot.dwFunction);
        single = TRUE;
    }

    fprintf(fp, "pci bus scan (using WD_PciConfigDump)\n");

    dwStatus = WDC_DriverOpen(WDC_DRV_OPEN_CHECK_VER, NULL);
    if (dwStatus)
        goto Exit;

    if (single)
        WDC_DIAG_PciDeviceInfoPrintFile(&pciSlot, fp, TRUE);
    else
        WDC_DIAG_PciDevicesInfoPrintAllFile(fp, TRUE);

Exit:
    WDC_DriverClose();
    if (dwStatus)
    {
        fprintf(fp, "%s failed, 0x%x - %s\n", argv[0], dwStatus,
            Stat2Str(dwStatus));
    }

    if (fp)
        fclose(fp);
    return dwStatus == WD_STATUS_SUCCESS ? 0 : -1;
}

File: samples/python/pci_dump/pci_dump.py
''' Jungo Connectivity Confidential. Copyright (c) 2022 Jungo Connectivity Ltd.  https://www.jungo.com '''

import sys, os
sys.path.append(os.path.dirname(os.path.abspath(__file__)) + "/..")
from wdlib.wdc_diag_lib import *
import argparse

def auto_int(x):
    return int(x, 0)

parser = argparse.ArgumentParser(description='PCI dump')

parser.add_argument('-f', '--filename', type=str)
parser.add_argument('-b', '--bus', type=auto_int)
parser.add_argument('-s', '--slot', type=auto_int)
parser.add_argument('-fu', '--function', type=auto_int)
args = parser.parse_args()

single = (args.bus != None and args.slot != None and args.function != None)

if not single:
    args.bus = 0
    args.slot = 0
    args.function = 0

if args.filename:
    fp = open(args.filename, "w+")
else:
    fp = None

fprintf(fp, "PCI dump (using WD_PciConfigDump)\n")

dwStatus = wdapi.WDC_DriverOpen(WDC_DRV_OPEN_CHECK_VER, 0)
if dwStatus:
    print("WDC_DriverOpen: " + Stat2Str(dwStatus))
    exit

if single:
    pSlot = WD_PCI_SLOT(args.bus, args.slot, args.function)
    WDC_DIAG_PciDeviceInfoPrintFile(pSlot, fp, True)
else:
    WDC_DIAG_PciDevicesInfoPrintAllFile(fp, True)

dwStatus = wdapi.WDC_DriverClose()
if dwStatus:
    print("WDC_DriverClose: " + Stat2Str(dwStatus))

if fp:
    fp.close()
